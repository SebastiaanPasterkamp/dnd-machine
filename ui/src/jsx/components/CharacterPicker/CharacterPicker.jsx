import React from 'react';
import PropTypes from 'prop-types';

import BaseFilterListDialog from '../BaseFilterListDialog';
import CharacterRow from './components/CharacterRow.jsx';
import InputField from '../InputField.jsx';
import ObjectDataListWrapper from '../../hocs/ObjectDataListWrapper.jsx';

import './sass/_character-picker.scss';


export class CharacterPicker extends React.Component
{
    constructor(props) {
        super(props);
        this.state = {
            currentPick: null,
            search: '',
        };
        this.onDone = this.onDone.bind(this);
        this.onFilter = this.onFilter.bind(this);
        this.onPick = this.onPick.bind(this);
        this.onSearch = this.onSearch.bind(this);
    }

    onDone() {
        const { onDone } = this.props;
        const { currentPick } = this.state;
        onDone(currentPick);
    }

    onFilter(character) {
        const { onFilter } = this.props;
        const { search } = this.state;
        if (onFilter !== undefined
            && !onFilter(character, search)
        ) {
            return false;
        }

        if (!search) {
            return true;
        }

        const pattern = search.toLowerCase();
        return (
            character.name.toLowerCase().match(pattern)
            || character.class.toLowerCase().match(pattern)
            || character.race.toLowerCase().match(pattern)
            || character.background.toLowerCase().match(pattern)
        );
    }

    onPick(currentPick) {
        this.setState({ currentPick });
    }

    onSearch(search) {
        this.setState({
            search: search || '',
        });
    }

    render() {
        const {
            characters,
            onCancel,
            onDone,
            onFilter,
            ...props
        } = this.props
        const { currentPick, search } = this.state;

        return (
            <BaseFilterListDialog
                items={characters}
                label="Pick a Character"
                className="character-picker"
                rowComponent={CharacterRow}
                onCancel={onCancel}
                onDone={this.onDone}
                onFilter={this.onFilter}
                onPick={this.onPick}
                currentPick={currentPick}
                {...props}
            >
                <InputField
                    placeholder="Filter..."
                    value={search}
                    setState={this.onSearch}
                />
            </BaseFilterListDialog>
        );
    }
};

CharacterPicker.propTypes = {
    onDone: PropTypes.func.isRequired,
    onFilter: PropTypes.func,
    onCancel: PropTypes.func,
    characters: PropTypes.object,
};

CharacterPicker.defaultProps = {
    characters: {},
    onCancel: null,
};

export default ObjectDataListWrapper(
    CharacterPicker,
    {characters: {type: 'character'}}
);
