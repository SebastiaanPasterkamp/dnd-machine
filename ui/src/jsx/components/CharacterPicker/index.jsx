export { default, CharacterPicker } from './CharacterPicker';
