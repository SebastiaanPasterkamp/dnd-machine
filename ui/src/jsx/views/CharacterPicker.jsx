import React from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';

import '../../sass/_character-picker.scss';

import ObjectDataListWrapper from '../hocs/ObjectDataListWrapper.jsx';

import CharacterLabel from '../components/CharacterLabel.jsx';
import InputField from '../components/InputField.jsx';
import UserLabel from '../components/UserLabel.jsx';

export class CharacterPicker extends React.Component
{
    constructor(props) {
        super(props);
    }

    render() {
        const {
            actions, characters = [], filter, showUser
        } = this.props;

        const filtered = _.isFunction(filter)
            ? _.filter(
                characters,
                character => filter(character)
            )
            : characters;

        return <table
            className="nice-table condensed bordered character-picker"
            >
            <thead>
                <tr>
                    {showUser ? <th>Player</th> : null }
                    <th>Character</th>
                    <th>Action</th>
                </tr>
            </thead>
            <tbody>{_.map(filtered, character => (
                <tr key={character.id}>
                    {showUser ? <td>
                        <UserLabel
                            user_id={character.user_id}
                            />
                    </td> : null}
                    <td>
                        <CharacterLabel
                            character_id={character.id}
                            showProgress={true}
                            />
                    </td>
                    <td>
                        {actions(character)}
                    </td>
                </tr>
            ))}</tbody>
        </table>;
    }
}

CharacterPicker.propTypes = {
    actions: PropTypes.func.isRequired,
    characters: PropTypes.oneOfType([
        PropTypes.object,
        PropTypes.array
    ]),
    filter: PropTypes.func,
    showUser: PropTypes.bool,
};

export default ObjectDataListWrapper(
    CharacterPicker,
    {characters: {type: 'character'}}
);
