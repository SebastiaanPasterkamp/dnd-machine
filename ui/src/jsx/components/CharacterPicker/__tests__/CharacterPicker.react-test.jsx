import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import MockRouter from 'react-mock-router';

import CharacterPicker from '../CharacterPicker.jsx';

describe('CharacterPicker', () => {
    const fullProps = {
        onDone: jest.fn(),
        onFilter: jest.fn(c => c.user_id === 2),
        onCancel: jest.fn(),
        label: "Renamed dialog",
        characters: {
            1: {
                id: 1,
                user_id: 2,
                race: 'Human',
                class: 'Fighter',
                background: 'Soldier',
                name: 'Testy McTestface',
                level: 1,
                gender: 'male',
                alignment: 'true neutral',
                adventure_checkpoints: 3,
                acp_progress: 3,
                acp_level: 4,
            },
            2: {
                id: 2,
                user_id: 1,
                race: 'Tiefling',
                class: 'Rogue',
                name: 'Stabby McStabface',
            },
            3: {
                id: 3,
                user_id: 2,
                race: 'Half-Orc',
                class: 'Druid',
                background: 'Sage',
                name: 'Beasty McBeastface',
                level: 2,
                gender: 'female',
                alignment: 'neutral good',
                xp: 900,
                xp_progress: 500,
                xp_level: 600,
            },
        },
    };

    describe('rendering', () => {
        it('should work with minimum props', () => {
            const tree = renderer.create(
                <MockRouter>
                    <CharacterPicker
                        onDone={fullProps.onDone}
                        onFilter={fullProps.onFilter}
                    />
                </MockRouter>
            );

            expect(tree).toMatchSnapshot();
        });

        it('should work with full props', () => {
            const tree = renderer.create(
                <MockRouter>
                    <CharacterPicker
                        {...fullProps}
                    />
                </MockRouter>
            );

            expect(tree).toMatchSnapshot();
        });
    });

    describe('filtering', () => {
        const wrapper = mount(
            <MockRouter>
                <CharacterPicker
                    {...fullProps}
                />
            </MockRouter>
        );

        afterEach(() => jest.clearAllMocks());

        it('should use provided filter', () => {
            expect(wrapper.text())
                .toMatch(/Testy McTestface/);
            expect(wrapper.text())
                .not
                .toMatch(/Stabby McStabface/);
            expect(wrapper.text())
                .toMatch(/Beasty McBeastface/);
        });

        it('should use text filter', () => {
            wrapper
                .find('input')
                .simulate('change', {target: {value: 'beast'}});
            expect(wrapper.text())
                .not
                .toMatch(/Testy McTestface/);
            expect(wrapper.text())
                .not
                .toMatch(/Stabby McStabface/);
            expect(wrapper.text())
                .toMatch(/Beasty McBeastface/);
        });

        it('should use clear filter', () => {
            wrapper
                .find('input')
                .simulate('change', {target: {value: ''}});
            expect(wrapper.text())
                .toMatch(/Testy McTestface/);
            expect(wrapper.text())
                .not
                .toMatch(/Stabby McStabface/);
            expect(wrapper.text())
                .toMatch(/Beasty McBeastface/);
        });
    });

    describe('allows picking a Character', () => {
        const wrapper = mount(
            <MockRouter>
                <CharacterPicker
                    {...fullProps}
                />
            </MockRouter>
        );

        afterEach(() => jest.clearAllMocks());

        it('should remember the selected character', () => {
            expect(wrapper.exists('a.accent')).toBe(false);
            wrapper
                .find('.fa-user-secret')
                .at(0)
                .simulate('click');
            expect(wrapper.exists('a.accent')).toBe(true);
        });

        it('should emit the current selection', () => {
            wrapper
                .find('.fa-check')
                .simulate('click');

            expect(fullProps.onDone)
                .toBeCalledWith(fullProps.characters[1].id);
        });

        it('should allow canceling', () => {
            wrapper
                .find('.fa-cross')
                .simulate('click');

            expect(fullProps.onCancel)
                .toBeCalled();
            expect(fullProps.onDone)
                .not
                .toBeCalled();
        });
    });
});
