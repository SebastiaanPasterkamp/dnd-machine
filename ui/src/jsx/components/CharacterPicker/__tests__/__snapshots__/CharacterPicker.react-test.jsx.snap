// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterPicker rendering should work with full props 1`] = `
<div
  className="modal-dialog nice-modal viewport-center accent wide"
>
  <div
    className="nice-modal-content"
  >
    <div
      className="nice-modal-header"
    >
      <a
        className="nice-modal-close"
        onClick={[Function]}
      >
        <i
          className="icon fa-times"
        />
      </a>
      <h4>
        Renamed dialog
      </h4>
    </div>
    <div
      className="nice-modal-sub"
    >
      <input
        className="nice-form-control"
        disabled={undefined}
        onChange={[Function]}
        onKeyPress={null}
        placeholder="Filter..."
        type={undefined}
        value=""
      />
    </div>
    <div
      className="nice-modal-body overflow-y-auto overflow-x-hidden"
    >
      <div
        className="character-picker__row"
      >
        <div
          className="character inline"
        >
          <div
            className="character-label"
          >
            <span
              className="character-label__name"
            >
              Testy McTestface
            </span>
            <span
              className="character-label__item"
            >
              Level 
              1
            </span>
            <div
              className="list-label inline"
            >
              male
            </div>
            <span
              className="character-label__item"
            >
              Human
            </span>
            <span
              className="character-label__item"
            >
              Fighter
            </span>
            <div
              className="list-label inline character-label__alignment"
            >
              true neutral
            </div>
          </div>
          <div
            className="nice-progress brand"
          >
            <div
              className="nice-progress-fill"
              style={
                Object {
                  "height": "1.25rem",
                  "width": "75%",
                }
              }
            >
              Level 1 (3 / 4)
            </div>
          </div>
        </div>
        <div
          className="actions"
        >
          <div
            className="base-link-group nice-btn-group"
          >
            <a
              className="cursor-pointer nice-btn icon fa-user-secret"
              onClick={[Function]}
            >
              Pick
            </a>
          </div>
        </div>
      </div>
      <div
        className="character-picker__row"
      >
        <div
          className="character inline"
        >
          <div
            className="character-label"
          >
            <span
              className="character-label__name"
            >
              Beasty McBeastface
            </span>
            <span
              className="character-label__item"
            >
              Level 
              2
            </span>
            <div
              className="list-label inline"
            >
              female
            </div>
            <span
              className="character-label__item"
            >
              Half-Orc
            </span>
            <span
              className="character-label__item"
            >
              Druid
            </span>
            <div
              className="list-label inline character-label__alignment"
            >
              neutral good
            </div>
          </div>
          <div
            className="nice-progress good"
          >
            <div
              className="nice-progress-fill"
              style={
                Object {
                  "height": "1.25rem",
                  "width": "83.33333333333334%",
                }
              }
            >
              Level 2 (500 / 600)
            </div>
          </div>
        </div>
        <div
          className="actions"
        >
          <div
            className="base-link-group nice-btn-group"
          >
            <a
              className="cursor-pointer nice-btn icon fa-user-secret"
              onClick={[Function]}
            >
              Pick
            </a>
          </div>
        </div>
      </div>
    </div>
    <div
      className="nice-modal-footer"
    >
      <div
        className="base-link-group nice-btn-group pull-right"
      >
        <a
          className="cursor-pointer nice-btn icon fa-cross"
          onClick={[Function]}
        >
          Cancel
        </a>
        <a
          className="cursor-pointer nice-btn icon fa-check primary"
          onClick={[Function]}
        >
          Done
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`CharacterPicker rendering should work with minimum props 1`] = `
<div
  className="modal-dialog nice-modal viewport-center accent wide"
>
  <div
    className="nice-modal-content"
  >
    <div
      className="nice-modal-header"
    >
      <h4>
        Pick a Character
      </h4>
    </div>
    <div
      className="nice-modal-sub"
    >
      <input
        className="nice-form-control"
        disabled={undefined}
        onChange={[Function]}
        onKeyPress={null}
        placeholder="Filter..."
        type={undefined}
        value=""
      />
    </div>
    <div
      className="nice-modal-body overflow-y-auto overflow-x-hidden"
    />
    <div
      className="nice-modal-footer"
    >
      <div
        className="base-link-group nice-btn-group pull-right"
      >
        <a
          className="cursor-pointer nice-btn icon fa-check primary"
          onClick={[Function]}
        >
          Done
        </a>
      </div>
    </div>
  </div>
</div>
`;
